syntax = "proto3";

option go_package = "./go_client";

package helloworld;

// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {}
  // Computes mean
  rpc ComputeMean (stream IntNumber) returns (FloatNumber) {}

  rpc ComputeMeanStream (stream IntNumber) returns (stream FloatNumber) {}

  rpc SendLongString (LongString) returns (LongString) {}

  rpc ComputeMeanOrSendLongString (stream FloatOrLongString) returns (FloatOrLongString) {}
}

// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
}

message IntNumber {
  int32 value = 1;
}

message FloatNumber {
  float value = 1;
}

message LongString {
  bytes str = 1;
}

message FloatOrLongString {
  oneof type {
    FloatNumber float_number_type = 1;
    LongString long_string_type = 2;
  }
}
